File: NeetCode 150/DP/152. Maximum Product Subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maxprod = nums[0]
        currmax = nums[0]
        currmin = nums[0]

        for i in range(1,len(nums)):
            temp = max(nums[i] * currmax,nums[i]*currmin,nums[i])
            currmin = min(nums[i] * currmax,nums[i]*currmin,nums[i])
            currmax = temp
            maxprod = max(maxprod,currmax)
        return maxprod

File: NeedCode 150 (JAVA)/DP/152. Maximum Product Subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int maxProd = nums[0];
        int currMin = nums[0];
        int currMax = nums[0];

        for(int i = 1 ; i<nums.length;i++){
            int temp = Math.max(nums[i], Math.max(nums[i] * currMin,nums[i] * currMax));

            currMin = Math.min(nums[i] , Math.min(nums[i] * currMin, nums[i] * currMax));

            currMax = temp;
            maxProd = Math.max(maxProd, currMax);
        }

        return maxProd;
    }
}
